### PR TITLE
Avrohugger config changes and Scalafix rewrites to make Avrohugger's output work in Scala 3

### PR DESCRIPTION
**Pull request by cb372**
A few changes to Avrohugger's generated sources to make them work in Scala 3 with avro4s v5.0.0.M1.

## Avro enums

Previously, we used Avrohugger's default configuration, which generates Scala 2 enumerations. Those are still valid in Scala 3, but avro4s 5.0.0.M1 doesn't support them.

I've configured Avrohugger to generate a sealed trait + case objects instead, which avro4s correctly handles as an Avro enum. Also added a Scalafix rule to add an avro4s annotation to the case objects so the correct order of enum values is preserved.

However, avro4s 4.x, which we use for Scala 2, doesn't support the encoding of Avro enums as sealed trait + case objects. So for Scala 2, we configure Avrohugger to generate Scala 2 enumerations just like we have up to now.

To summarise:

* Scala 2 / avro4s 4.x
	* keep the same Avrohugger configuration as before, generate a Scala 2 enumeration 
* Scala 3 / avro4s 5.x
    * configure Avrohugger to generate a sealed trait + case objects
    * Scalafix rule to add the `@AvroSortPriority` annotation

## Avro unions

For unions of 3 or more types, Avrohugger generates a shapeless Coproduct, e.g. `A :+: B :+: C :+: CNil`. We can't use that in Scala 3, so for Scala 3 we apply a Scalafix rule to rewrite the type to `higherkindness.mu.rpc.avro.AvroUnion3[A, B, C]`.

That type is a wrapper around a Scala 3 union type `A | B | C`, with a custom avro4s codec to encode it as an Avro union. It's defined in https://github.com/higherkindness/mu-scala/pull/1453. I've implemented support for unions of up to 8 types.

For Scala 2, we don't have Scala union types, so we continue to use shapeless Coproduct.

## Decimals

Avrohugger supports encoding Avro decimals either as plain `BigDecimal` or as a `BigDecimal` tagged with shapeless `Nat`s to store the precision and scale.

~~For now I've simply removed this feature, hardcoding Avrohugger's configuration to generate plain `BigDecimal`.~~

~~If we need to resurrect this feature in the future, it should be possible to build our own tagging mechanism using an opaque type and Scala 3 literal types. However, Avro4s provides a separate mechanism whereby users can specify the scale and precision of BigDecimals by providing an implicit `ScalePrecision`, so I don't think this feature is needed anymore.~~

I've added a Scalafix rule to rewrite the shapeless-tagged types (e.g. `BigDecimal @@ (Nat._6, Nat._2)`) to the equivalent Mu `TaggedDecimal` type (e.g. `TaggedDecimal[6, 2]`). This rule is only applied when generating Scala 3 code.

See https://github.com/higherkindness/mu-scala/pull/1457 for details of the `TaggedDecimal` type.

### CODE REVIEW

**Comment by fedefernandez** [comment]
@cb372 this looks great!

My main concern is related to the precision and the scale. We worked on adding this feature because we have some use cases on clients where the same service had to support different precisions and scales, making the use of the implicit `ScalePrecision` not suitable. With the tagged mechanism, we provided support to specify that in the avdl files.

https://github.com/julianpeeters/avrohugger/pull/102

I think it's a feature we would need to keep.

**Comment by cb372** [comment]
OK, I'll spend a bit of time digging into how we can support that feature in Scala 3.

**Review by fedefernandez — APPROVED** [review]
Code looks good, we could address that feature in a following PR

**Comment by cb372** [comment]
I've added support for tagged BigDecimals, so this is now ready for another review.

File: plugin/src/main/scala/higherkindness/mu/rpc/srcgen/SrcGenPlugin.scala [hunk 1]
```diff
@@ -135,11 +135,7 @@ object SrcGenPlugin extends AutoPlugin {
     muSrcGenMarshallerImports := {
       muSrcGenSerializationType.value match {
         case SerializationType.Avro | SerializationType.AvroWithSchema =>
-          val bigDecimal = muSrcGenBigDecimal.value match {
-            case ScalaBigDecimalGen       => BigDecimalAvroMarshallers
-            case ScalaBigDecimalTaggedGen => BigDecimalTaggedAvroMarshallers
-          }
-          List(bigDecimal, JavaTimeDateAvroMarshallers)
```
Comment (fedefernandez) [inline plugin/src/main/scala/higherkindness/mu/rpc/srcgen/SrcGenPlugin.scala:142]: I don't fully understand why we don't need the imports for big decimal marshallers

**Review by fedefernandez — APPROVED** [review]
Code looks good, thanks @cb372. I've left a question

File: plugin/src/main/scala/higherkindness/mu/rpc/srcgen/SrcGenPlugin.scala (hunk 1 above)
Comment (cb372) [reply plugin/src/main/scala/higherkindness/mu/rpc/srcgen/SrcGenPlugin.scala:142]: Good question. I'm not sure why I made this particular change - I think it's a mistake. But I'm planning to come back to the custom marshallers in a later PR.

I'll merge this as-is for now and revert this change in my next PR.

**Pull request merged**